FIX: Add a logger instance to use prior to starting ultravisor.

fitzy101 opened the pull request:
When configuring a service to use Logstash, we configure a LogstashWriter
instance to run via ultravisor. The @config.logger is then updated to
use the LogstashWriter. The problem with that, is if we then attempt to
use the @config.logger prior to calling `ultravisor#run` we get a thread
deadlock exception because the writer is not yet initialised.

In order to fix this race condition, create a @config.pre_logger
that is configured the same way as the standard logger, but is never
augmented after creation to use logstash writer (or other filters). That
way logs can be printed safely using pre_logger prior to the calls to
`ultravisor#run`.

The specific error message fixed is:

```
/usr/local/bundle/gems/ultravisor-0.0.0.3.g8cf10dc/lib/ultravisor/child.rb:400:in
`sleep': No live threads left. Deadlock? (fatal)
1 threads, 1 sleeps current:0x000055ed5fa79540 main
  thread:0x000055ed5fa79540
* #<Thread:0x000055ed5faacb78 sleep_forever>
```
because a child instance is attempted to be accessed before it is
started.

fitzy101 committed:
FIX: Add a logger instance to use prior to starting ultravisor.

When configuring a service to use Logstash, we configure a LogstashWriter
instance to run via ultravisor. The @config.logger is then updated to
use the LogstashWriter. The problem with that, is if we then attempt to
use the @config.logger prior to calling `ultravisor#run` we get a thread
deadlock exception because the writer is not yet initialised.

In order to fix this race condition, create a @config.pre_logger
that is configured the same way as the standard logger, but is never
augmented after creation to use logstash writer (or other filters). That
way logs can be printed safely using pre_logger prior to the calls to
`ultravisor#run`.

The specific error message fixed is:

```
/usr/local/bundle/gems/ultravisor-0.0.0.3.g8cf10dc/lib/ultravisor/child.rb:400:in
`sleep': No live threads left. Deadlock? (fatal)
1 threads, 1 sleeps current:0x000055ed5fa79540 main
  thread:0x000055ed5fa79540
* #<Thread:0x000055ed5faacb78 sleep_forever>
```
because a child instance is attempted to be accessed before it is
started.

martin-brennan reviewed (approved):
I mean...this _seems_ fine to me but I am not really across what logstash is doing. Both loggers are writing to the same file so I suppose the logs will be picked up OK?

Does this call to the logger happen _after_ ultravisor.run? https://github.com/discourse/service_skeleton/blob/d9eba057f7a1ecef4cb3c8accaa49cc7ee900554/lib/service_skeleton/generator.rb#L65

Maybe `unsupervised_logger` would be a nicer name than `pre_logger` as well?

fitzy101 commented:
By default everything logs to STDOUT, and can be configured otherwise to write to a file. Logstash is written to asynchronously as it has to cross a network boundary, with the async log writer being launched by ultravisor, and isn't created at all if the LOGSTASH_SERVER env isn't set.

The logs in `generator.rb` are prior to the ultravisor run - generator is used in the initialisation stage prior to the `ServiceSkeleton::Runner#run` call and can therefore be updated to use the 'pre' runner. I'm also not happy with that name but not sure if `unsupervised_logger` is any better... what about `init_logger` as it will only ever be used during the initialisation stage?

martin-brennan commented:
@fitzy101 thanks for the explanation!

> what about init_logger as it will only ever be used during the initialisation stage?

Yeah that is better, or even `pre_ultravisor_logger`. Either way is good, but yeah init is better than just pre.

fitzy101 committed:
Update name of logger to indicate when it should be used.

Also change the other logs emitted during the config generation stage to
use the pre_run_logger.

fitzy101 committed:
Fix failing specs due to creating 2 loggers.
